File: packages/adapters/src/marginalia_adapters/fake/playback.py
"""Fake playback engine."""

from __future__ import annotations

from marginalia_core.domain.document import Document
from marginalia_core.domain.reading_session import PlaybackState, ReadingPosition
from marginalia_core.ports.capabilities import ProviderCapabilities
from marginalia_core.ports.playback import PlaybackSnapshot
from marginalia_core.ports.tts import SynthesisResult

PLAYBACK_CAPABILITIES = ProviderCapabilities(
    provider_name="fake-playback",
    interface_kind="playback",
    supported_languages=("en",),
    supports_streaming=False,
    supports_partial_results=False,
    supports_timestamps=True,
    low_latency_suitable=True,
    offline_capable=True,
)


class FakePlaybackEngine:
    """Track playback commands without attempting real audio output."""

    def __init__(self, *, auto_complete_after_snapshots: int | None = None) -> None:
        self.state = PlaybackState.STOPPED
        self.last_document_id: str | None = None
        self.last_position = ReadingPosition()
        self.last_action = "stopped"
        self.progress_units = 0
        self.audio_reference: str | None = None
        self._auto_complete_after_snapshots = auto_complete_after_snapshots
        self._remaining_playing_snapshots: int | None = None

    def describe_capabilities(self) -> ProviderCapabilities:
        return PLAYBACK_CAPABILITIES

    def hydrate(self, snapshot: PlaybackSnapshot | None) -> None:
        if snapshot is None:
            return
        self.state = snapshot.state
        self.last_document_id = snapshot.document_id
        if snapshot.anchor is not None:
            self.last_position = _position_from_anchor(snapshot.anchor)
        self.last_action = snapshot.last_action
        self.progress_units = snapshot.progress_units
        self.audio_reference = snapshot.audio_reference

    def start(
        self,
        document: Document,
        position: ReadingPosition,
        *,
        synthesis: SynthesisResult | None = None,
    ) -> PlaybackSnapshot:
        self.last_document_id = document.document_id
        self.last_position = position
        self.state = PlaybackState.PLAYING
        self.last_action = "start"
        self.progress_units += 1
        self.audio_reference = synthesis.audio_reference if synthesis else None
        self._remaining_playing_snapshots = self._auto_complete_after_snapshots
        return self.snapshot()

    def pause(self) -> PlaybackSnapshot:
        self.state = PlaybackState.PAUSED
        self.last_action = "pause"
        return self.snapshot()

    def resume(self) -> PlaybackSnapshot:
        self.state = PlaybackState.PLAYING
        self.last_action = "resume"
        self.progress_units += 1
        self._remaining_playing_snapshots = self._auto_complete_after_snapshots
        return self.snapshot()

    def stop(self) -> PlaybackSnapshot:
        self.state = PlaybackState.STOPPED
        self.last_action = "stop"
        self.audio_reference = None
        self._remaining_playing_snapshots = None
        return self.snapshot()

    def seek(self, position: ReadingPosition) -> PlaybackSnapshot:
        self.last_position = position
        self.last_action = "seek"
        self.progress_units += 1
        return self.snapshot()

    def snapshot(self) -> PlaybackSnapshot:
        if (
            self.state is PlaybackState.PLAYING
            and self._remaining_playing_snapshots is not None
        ):
            if self._remaining_playing_snapshots <= 0:
                self.state = PlaybackState.STOPPED
                self.last_action = "completed"
            else:
                self._remaining_playing_snapshots -= 1
        return PlaybackSnapshot(
            state=self.state,
            last_action=self.last_action,
            document_id=self.last_document_id,
            anchor=self.last_position.anchor,
            progress_units=self.progress_units,
            audio_reference=self.audio_reference,
            provider_name=PLAYBACK_CAPABILITIES.provider_name,
            process_id=None,
        )


def _position_from_anchor(anchor: str) -> ReadingPosition:
    section_index = 0
    chunk_index = 0
    for item in anchor.split("/"):
        key, _, raw_value = item.partition(":")
        if key == "section":
            section_index = int(raw_value)
        elif key == "chunk":
            chunk_index = int(raw_value)
    return ReadingPosition(section_index=section_index, chunk_index=chunk_index)

File: crates/marginalia-provider-fake/src/playback.rs
use marginalia_core::domain::{Document, PlaybackState, ReadingPosition};
use marginalia_core::ports::{
    PlaybackEngine, PlaybackSnapshot, ProviderCapabilities, ProviderExecutionMode, SynthesisResult,
};

#[derive(Debug, Clone)]
pub struct FakePlaybackEngine {
    snapshot: PlaybackSnapshot,
}

impl Default for FakePlaybackEngine {
    fn default() -> Self {
        Self {
            snapshot: PlaybackSnapshot {
                state: PlaybackState::Stopped,
                last_action: "initialized".to_string(),
                document_id: None,
                anchor: None,
                progress_units: 0,
                audio_reference: None,
                provider_name: Some("fake-playback".to_string()),
                process_id: None,
            },
        }
    }
}

impl FakePlaybackEngine {
    pub fn new() -> Self {
        Self::default()
    }
}

impl PlaybackEngine for FakePlaybackEngine {
    fn describe_capabilities(&self) -> ProviderCapabilities {
        ProviderCapabilities {
            provider_name: "fake-playback".to_string(),
            interface_kind: "playback".to_string(),
            supported_languages: vec!["it".to_string(), "en".to_string()],
            supports_streaming: false,
            supports_partial_results: false,
            supports_timestamps: false,
            low_latency_suitable: true,
            offline_capable: true,
            execution_mode: ProviderExecutionMode::Local,
        }
    }

    fn hydrate(&mut self, snapshot: Option<PlaybackSnapshot>) {
        if let Some(snapshot) = snapshot {
            self.snapshot = snapshot;
        } else {
            self.snapshot.state = PlaybackState::Stopped;
            self.snapshot.last_action = "hydrated-empty".to_string();
            self.snapshot.document_id = None;
            self.snapshot.anchor = None;
            self.snapshot.audio_reference = None;
            self.snapshot.process_id = None;
        }
    }

    fn start(
        &mut self,
        document: &Document,
        position: &ReadingPosition,
        synthesis: Option<SynthesisResult>,
    ) -> PlaybackSnapshot {
        self.snapshot.state = PlaybackState::Playing;
        self.snapshot.last_action = "start".to_string();
        self.snapshot.document_id = Some(document.document_id.clone());
        self.snapshot.anchor = Some(position.anchor());
        self.snapshot.progress_units = position.chunk_index;
        self.snapshot.audio_reference = synthesis.map(|synthesis| synthesis.audio_reference);
        self.snapshot.clone()
    }

    fn pause(&mut self) -> PlaybackSnapshot {
        self.snapshot.state = PlaybackState::Paused;
        self.snapshot.last_action = "pause".to_string();
        self.snapshot.clone()
    }

    fn resume(&mut self) -> PlaybackSnapshot {
        self.snapshot.state = PlaybackState::Playing;
        self.snapshot.last_action = "resume".to_string();
        self.snapshot.clone()
    }

    fn stop(&mut self) -> PlaybackSnapshot {
        self.snapshot.state = PlaybackState::Stopped;
        self.snapshot.last_action = "stop".to_string();
        self.snapshot.clone()
    }

    fn seek(&mut self, position: &ReadingPosition) -> PlaybackSnapshot {
        self.snapshot.anchor = Some(position.anchor());
        self.snapshot.progress_units = position.chunk_index;
        self.snapshot.last_action = "seek".to_string();
        self.snapshot.clone()
    }

    fn snapshot(&self) -> PlaybackSnapshot {
        self.snapshot.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::FakePlaybackEngine;
    use marginalia_core::domain::{
        Document, DocumentChunk, DocumentSection, PlaybackState, ReadingPosition,
    };
    use marginalia_core::ports::{PlaybackEngine, SynthesisResult};
    use std::collections::HashMap;
    use std::path::PathBuf;

    #[test]
    fn fake_playback_engine_updates_snapshot_during_controls() {
        let mut engine = FakePlaybackEngine::new();
        let document = Document {
            document_id: "doc-1".to_string(),
            title: "Doc".to_string(),
            source_path: PathBuf::from("/tmp/doc.md"),
            sections: vec![DocumentSection {
                index: 0,
                title: "Intro".to_string(),
                chunks: vec![DocumentChunk {
                    index: 0,
                    text: "Alpha".to_string(),
                    char_start: 0,
                    char_end: 5,
                }],
                source_anchor: Some("section:0".to_string()),
            }],
            imported_at: chrono::Utc::now(),
        };
        let position = ReadingPosition::default();

        let started = engine.start(
            &document,
            &position,
            Some(SynthesisResult {
                provider_name: "fake-tts".to_string(),
                voice: "narrator".to_string(),
                content_type: "audio/wav".to_string(),
                audio_reference: "/tmp/audio.wav".to_string(),
                byte_length: 42,
                text_excerpt: "Alpha".to_string(),
                metadata: HashMap::new(),
            }),
        );
        assert_eq!(started.state, PlaybackState::Playing);

        let paused = engine.pause();
        assert_eq!(paused.state, PlaybackState::Paused);

        let stopped = engine.stop();
        assert_eq!(stopped.state, PlaybackState::Stopped);
    }
}
